File: COM Scripts/C#.Net/COM_TestForm/COM_TestForm/Form1.cs
/*
 * PDFCreator COM samples for C#
 * Part of the PDFCreator application
 * License: GPL
 * Homepage: http://www.pdfforge.org/pdfcreator
 * .Net Framework: 4.7.2
 * Version: 1.1.0.0
 * Created: May, 25. 2020
 * Modified: May, 25. 2020
 * Author: pdfforge GmbH
 * Comments: This project demonstrates the use of the COM Interface of PDFCreator.
             There are 5 different kinds of usage presented.
             Further usage presentation is only available in the JavaScript directory.
 * Note: When executed in release mode paths have to be modified.
 */

using pdfforge.PDFCreator.UI.ComWrapper;
using System;
using System.Data;
using System.IO;
using System.Reflection;
using System.Windows.Forms;

using pdfforge.PDFCreator.UI.ComWrapper;
//IMPORTANT: Add a reference to the PDFCreator.ComWrapper.dll

namespace COM_TestForm
{
    public partial class Form1 : Form
    {
        private bool _isTypeInitialized;

        public Form1()
        {
            InitializeComponent();
            UpdateStatus("Program started");
        }

        private JobQueue CreateQueue()
        {
            // This needs to be done once to make the ComWrapper work reliably.
            if (!_isTypeInitialized)
            {
                Type queueType = Type.GetTypeFromProgID("PDFCreator.JobQueue");
                Activator.CreateInstance(queueType);
                _isTypeInitialized = true;
            }

            return new JobQueue();
        }

        private void testPage_btn_Click(object sender, EventArgs e)
        {
            const string sampleName = "TestPage2PDF";

            UpdateStatus(string.Format(Messages.SampleStarted, sampleName));

            var captionName = Messages.DefaultMessageCaption + ": " + sampleName;
            var convertedFilePath = Path.Combine(Path.GetTempPath(), sampleName + ".pdf");

            var jobQueue = CreateQueue();

            try
            {
                MessageBox.Show(Messages.Initializing, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.Initialize();

                MessageBox.Show(Messages.PrintingTestPage, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                PrintWindowsTestPage();

                const int waitTimeSeconds = 10;
                if (!jobQueue.WaitForJob(waitTimeSeconds))
                {
                    MessageBox.Show(string.Format(Messages.JobDidNotArrive, waitTimeSeconds), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                }
                else
                {
                    MessageBox.Show(string.Format(Messages.JobsInQueue, jobQueue.Count), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    MessageBox.Show(Messages.GettingJobInstance, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    var printJob = jobQueue.NextJob;

                    const string profile = "DefaultGuid";
                    printJob.SetProfileByGuidOrName(profile);

                    MessageBox.Show(string.Format(Messages.ConvertingWithProfile, profile), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    printJob.ConvertTo(convertedFilePath);

                    if (!printJob.IsFinished || !printJob.IsSuccessful)
                    {
                        UpdateStatus(Messages.CouldNotConvert);
                        MessageBox.Show(Messages.CouldNotConvert, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                    else
                    {
                        UpdateStatus(string.Format(Messages.ConversionSuccessfulFile1, convertedFilePath));
                        MessageBox.Show(string.Format(Messages.ConversionSuccessfulFile2, convertedFilePath), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                }
            }
            catch (Exception err)
            {
                MessageBox.Show(string.Format(Messages.Error, err.Message), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
            finally
            {
                MessageBox.Show(Messages.ReleasingQueueObject, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.ReleaseCom();
            }
        }

        private void PrintWindowsTestPage()
        {
            Type shellObj = Type.GetTypeFromProgID("Shell.Application");
            dynamic shellInst = Activator.CreateInstance(shellObj);

            shellInst.ShellExecute("RUNDLL32.exe", "PRINTUI,PrintUIEntry /k /n \"PDFCreator\"", "", "open", 1);
        }

        private void jpegSettings_btn_Click(object sender, EventArgs e)
        {
            const string sampleName = "TestPage2Jpeg";

            UpdateStatus(string.Format(Messages.SampleStarted, sampleName));

            var captionName = Messages.DefaultMessageCaption + ": " + sampleName;
            var convertedFilePath = Path.Combine(Path.GetTempPath(), sampleName + ".jpg");

            var jobQueue = CreateQueue();

            try
            {
                MessageBox.Show(Messages.Initializing, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.Initialize();

                MessageBox.Show(Messages.PrintingTestPage, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                PrintWindowsTestPage();

                const int waitTimeSeconds = 10;
                if (!jobQueue.WaitForJob(waitTimeSeconds))
                {
                    MessageBox.Show(string.Format(Messages.JobDidNotArrive, waitTimeSeconds), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                }
                else
                {
                    MessageBox.Show(string.Format(Messages.JobsInQueue, jobQueue.Count), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    MessageBox.Show(Messages.GettingJobInstance, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    var printJob = jobQueue.NextJob;

                    const string profile = "JpegGuid";
                    printJob.SetProfileByGuidOrName(profile);

                    MessageBox.Show("Applying jpeg settings", captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    printJob.SetProfileSetting("JpegSettings.Color", "Color24Bit");
                    printJob.SetProfileSetting("JpegSettings.Quality", "100");

                    MessageBox.Show(string.Format(Messages.ConvertingWithProfile, profile), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    printJob.ConvertTo(convertedFilePath);

                    if (!printJob.IsFinished || !printJob.IsSuccessful)
                    {
                        UpdateStatus(Messages.CouldNotConvert);
                        MessageBox.Show(Messages.CouldNotConvert, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                    else
                    {
                        UpdateStatus(string.Format(Messages.ConversionSuccessfulFile1, convertedFilePath));
                        MessageBox.Show(string.Format(Messages.ConversionSuccessfulFile2, convertedFilePath), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                }
            }
            catch (Exception err)
            {
                MessageBox.Show(string.Format(Messages.Error, err.Message), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
            finally
            {
                MessageBox.Show(Messages.ReleasingQueueObject, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.ReleaseCom();
            }
        }

        private void mergedFiles_btn_Click(object sender, EventArgs e)
        {
            const string sampleName = "MergedMultipleFiles2Tif";

            UpdateStatus(string.Format(Messages.SampleStarted, sampleName));

            var captionName = Messages.DefaultMessageCaption + ": " + sampleName;
            var convertedFilePath = Path.Combine(Path.GetTempPath(), sampleName + ".tif");

            var jobQueue = CreateQueue();

            try
            {
                MessageBox.Show(Messages.Initializing, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.Initialize();

                var countTestPages = 3;
                MessageBox.Show(string.Format("Printing {0} windows test pages ...", countTestPages), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                for (int i = 0; i < countTestPages; i++)
                    PrintWindowsTestPage();

                const int waitTimeSeconds = 15;
                if (!jobQueue.WaitForJobs(countTestPages, waitTimeSeconds))
                {
                    MessageBox.Show(string.Format(Messages.JobDidNotArrive, waitTimeSeconds), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                }
                else
                {
                    MessageBox.Show(string.Format(Messages.JobsInQueue, jobQueue.Count), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    MessageBox.Show("Merging all available jobs", captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    jobQueue.MergeAllJobs();

                    MessageBox.Show(string.Format(Messages.JobsInQueue, jobQueue.Count), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    MessageBox.Show(Messages.GettingJobInstance, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    var printJob = jobQueue.NextJob;

                    const string profile = "DefaultGuid";
                    printJob.SetProfileByGuidOrName(profile);

                    //Change the output format of the current conversion profile
                    //to .tif - this holds only for this job
                    printJob.SetProfileSetting("OutputFormat", "Tif");

                    MessageBox.Show(string.Format("Converting with profile \"{0}\" but with .tif as output format.", profile), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    printJob.ConvertTo(convertedFilePath);

                    if (!printJob.IsFinished || !printJob.IsSuccessful)
                    {
                        UpdateStatus(Messages.CouldNotConvert);
                        MessageBox.Show(Messages.CouldNotConvert, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                    else
                    {
                        UpdateStatus(string.Format(Messages.ConversionSuccessfulFile1, convertedFilePath));
                        MessageBox.Show(string.Format(Messages.ConversionSuccessfulFile2, convertedFilePath), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                }
            }
            catch (Exception err)
            {
                MessageBox.Show(string.Format(Messages.Error, err.Message), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
            finally
            {
                MessageBox.Show(Messages.ReleasingQueueObject, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.ReleaseCom();
            }
        }

        private void coverPage_btn_Click(object sender, EventArgs e)
        {
            const string sampleName = "CoverPage";

            UpdateStatus(string.Format(Messages.SampleStarted, sampleName));

            var captionName = Messages.DefaultMessageCaption + ": " + sampleName;
            var convertedFilePath = Path.Combine(Path.GetTempPath(), sampleName + ".pdf");

            var jobQueue = CreateQueue();

            var assemblyDir = Assembly.GetExecutingAssembly().Location;
            var coverPagePath = assemblyDir.Replace("\\bin\\Debug\\COM_TestForm.exe", "\\FilesForTests\\CoverPage.pdf");

            try
            {
                MessageBox.Show(Messages.Initializing, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.Initialize();

                MessageBox.Show(Messages.PrintingTestPage, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                PrintWindowsTestPage();

                const int waitTimeSeconds = 10;
                if (!jobQueue.WaitForJob(waitTimeSeconds))
                {
                    MessageBox.Show(string.Format(Messages.JobDidNotArrive, waitTimeSeconds), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                }
                else
                {
                    MessageBox.Show(string.Format(Messages.JobsInQueue, jobQueue.Count), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    MessageBox.Show(Messages.GettingJobInstance, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    var printJob = jobQueue.NextJob;

                    const string profile = "DefaultGuid";
                    printJob.SetProfileByGuidOrName(profile);
                    MessageBox.Show("Applying cover page settings", captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    printJob.AddAction("CoverPage");
                    printJob.SetProfileSetting("CoverPage.File", coverPagePath);

                    MessageBox.Show(string.Format(Messages.ConvertingWithProfile, profile), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    printJob.ConvertTo(convertedFilePath);

                    if (!printJob.IsFinished || !printJob.IsSuccessful)
                    {
                        UpdateStatus(Messages.CouldNotConvert);
                        MessageBox.Show(Messages.CouldNotConvert, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                    else
                    {
                        UpdateStatus(string.Format(Messages.ConversionSuccessfulFile1, convertedFilePath));
                        MessageBox.Show(string.Format(Messages.ConversionSuccessfulFile2, convertedFilePath), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                }
            }
            catch (Exception err)
            {
                MessageBox.Show(string.Format(Messages.Error, err.Message), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
            finally
            {
                MessageBox.Show(Messages.ReleasingQueueObject, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.ReleaseCom();
            }
        }

        private void backgroundPage_btn_Click(object sender, EventArgs e)
        {
            const string sampleName = "BackgroundPage";

            UpdateStatus(string.Format(Messages.SampleStarted, sampleName));

            var captionName = Messages.DefaultMessageCaption + ": " + sampleName;
            var convertedFilePath = Path.Combine(Path.GetTempPath(), sampleName + ".pdf");

            var jobQueue = CreateQueue();

            var assemblyDir = Assembly.GetExecutingAssembly().Location;
            var bkgroundPagePath = assemblyDir.Replace("\\bin\\Debug\\COM_TestForm.exe", "\\FilesForTests\\BackgroundPage.pdf");

            try
            {
                MessageBox.Show(Messages.Initializing, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.Initialize();

                MessageBox.Show(Messages.PrintingTestPage, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                PrintWindowsTestPage();

                const int waitTimeSeconds = 10;
                if (!jobQueue.WaitForJob(waitTimeSeconds))
                {
                    MessageBox.Show(string.Format(Messages.JobDidNotArrive, waitTimeSeconds), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                }
                else
                {
                    MessageBox.Show(string.Format(Messages.JobsInQueue, jobQueue.Count), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    MessageBox.Show(Messages.GettingJobInstance, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    var printJob = jobQueue.NextJob;
                    var list = printJob.GetProfileListSetting("ActionOrder");
                    const string profile = "DefaultGuid";
                    printJob.SetProfileByGuidOrName(profile);

                    MessageBox.Show("Applying background page settings", captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    printJob.AddActionToPosition("BackgroundPage", 1);
                    list = printJob.GetProfileListSetting("ActionOrder");
                    printJob.SetProfileSetting("BackgroundPage.Repetition", "RepeatAllPages");
                    printJob.SetProfileSetting("BackgroundPage.File", bkgroundPagePath);

                    MessageBox.Show(string.Format(Messages.ConvertingWithProfile, profile), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    printJob.ConvertTo(convertedFilePath);

                    if (!printJob.IsFinished || !printJob.IsSuccessful)
                    {
                        UpdateStatus(Messages.CouldNotConvert);
                        MessageBox.Show(Messages.CouldNotConvert, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                    else
                    {
                        UpdateStatus(string.Format(Messages.ConversionSuccessfulFile1, convertedFilePath));
                        MessageBox.Show(string.Format(Messages.ConversionSuccessfulFile2, convertedFilePath), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                }
            }
            catch (Exception err)
            {
                MessageBox.Show(string.Format(Messages.Error, err.Message), captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
            finally
            {
                MessageBox.Show(Messages.ReleasingQueueObject, captionName, MessageBoxButtons.OK, MessageBoxIcon.Information);
                jobQueue.ReleaseCom();
            }
        }

        private void UpdateStatus(string status)
        {
            toolStripStatusLabel1.Text = status;
        }

        private void statusStrip1_Resize(object sender, EventArgs e)
        {
            toolStripStatusLabel1.Width = statusStrip1.Width;
        }
    }

    public static class Messages
    {
        public const string SampleStarted = "Sample \"{0}\" started";
        public const string Initializing = "Initializing the job queue";
        public const string PrintingTestPage = "Printing windows test page ...";
        public const string JobDidNotArrive = "The job didn't arrive within {0} seconds.";
        public const string JobsInQueue = "Currently there are {0} job(s) in the queue.";
        public const string GettingJobInstance = "Getting job instance";
        public const string ConvertingWithProfile = "Converting with profile \"{0}\"";
        public const string ConversionSuccessfulFile1 = "The conversion was succesful! [File: {0}]";
        public const string ConversionSuccessfulFile2 = "The conversion was succesful!\r\n\r\nThe file was saved to: {0}";
        public const string CouldNotConvert = "Could not convert!";
        public const string Error = "An error occured: {0}";
        public const string ReleasingQueueObject = "Releasing the queue object";

        public const string DefaultMessageCaption = "PDFCreator c# com sample";
    }
}

File: COM Scripts/VB.Net/COM_TestForm/COM_TestForm/Form1.vb
'PDFCreator COM Interface tests for C#
'Part of the PDFCreator application
'License: GPL
'Homepage: http://www.pdfforge.org/pdfcreator
'.Net Framework: 4.7.2
'Version: 1.1.0.0
'Created: May, 25. 2020
'Modified: May, 25. 2020
'Author: pdfforge GmbH
'Comments: This project demonstrates the use of the COM Interface of PDFCreator.
'           There are 5 different kinds of usage presented.
'           Further usage presentation is only available in the JavaScript directory.
'Note: When executed in release mode paths have to be modified.

' IMPORTANT: Add a reference to the pdfcreator.exe
Imports System.IO
Imports pdfforge.PDFCreator.UI.ComWrapper

Public Class Form1

    Private Sub testPage_btn_Click(sender As Object, e As EventArgs) Handles testPage_btn.Click
        Dim PDFCreatorQueue As JobQueue = New JobQueue()
        Dim job As Object
        Dim fullPath As String
        Const sampleName = "TestPage2PDF"

        Try
            fullPath = Path.Combine(Path.GetTempPath, sampleName + ".pdf")
            MsgBox("Initializing PDFCreator queue...")
            PDFCreatorQueue.Initialize()

            MsgBox("Printing windows test page...")
            printFile()

            If Not PDFCreatorQueue.WaitForJob(10) Then
                MsgBox("The print job did not reach the queue within " & " 10 seconds")
            Else
                MsgBox("Currently there are " & PDFCreatorQueue.Count & " job(s) in the queue")
                MsgBox("Getting job instance")
                job = PDFCreatorQueue.NextJob

                job.SetProfileByGuidOrName("DefaultGuid")

                MsgBox("Converting under ""DefaultGuid"" conversion profile")
                job.ConvertTo(fullPath)

                If (Not job.IsFinished Or Not job.IsSuccessful) Then
                    MsgBox("Could not convert the file.")
                Else
                    MsgBox("Job finished successfully." + Environment.NewLine + Environment.NewLine + "File saved to: " & fullPath)
                End If
            End If
        Catch Ex As Exception
            MessageBox.Show("Original error: " & Ex.Message)
        Finally
            MsgBox("Releasing the queue object")
            PDFCreatorQueue.ReleaseCom()
        End Try

    End Sub

    Private Sub jpegSettings_btn_Click(sender As Object, e As EventArgs) Handles jpegSettings_btn.Click
        Dim oJobQueue As Type = Type.GetTypeFromProgID("PDFCreator.JobQueue")
        Dim PDFCreatorQueue As Object = Activator.CreateInstance(oJobQueue)
        Dim PrintJob As Object
        Dim fullPath As String
        Const sampleName = "TestPage2Jpeg"

        Try
            fullPath = Path.Combine(Path.GetTempPath, sampleName + ".jpg")

            MsgBox("Initializing PDFCreator queue...")
            PDFCreatorQueue.Initialize()

            MsgBox("Printing windows test page...")
            printFile()

            If Not PDFCreatorQueue.WaitForJob(10) Then
                MsgBox("The print job did not reach the queue within " & " 10 seconds")
            Else
                MsgBox("Currently there are " & PDFCreatorQueue.Count & " job(s) in the queue")
                MsgBox("Getting job instance")
                PrintJob = PDFCreatorQueue.NextJob

                PrintJob.SetProfileByGuidOrName("JpegGuid") 'Notice that we changed the profile

                'The SetProfileSettings method allows us to change the JpegSettings of the job
                'We want 24 bit colors for our converted file
                PrintJob.SetProfileSetting("JpegSettings.Color", "Color24Bit")
                PrintJob.SetProfileSetting("JpegSettings.Quality", "100")

                MsgBox("Converting under ""JpegGuid"" conversion profile")
                PrintJob.ConvertTo(fullPath)

                If (Not PrintJob.IsFinished Or Not PrintJob.IsSuccessful) Then
                    MsgBox("Could not convert the file.")
                Else
                    MsgBox("Job finished successfully." + Environment.NewLine + Environment.NewLine + "File saved to: " & fullPath)
                End If
            End If
        Catch Ex As Exception
            MessageBox.Show("Original error: " & Ex.Message)
        Finally
            MsgBox("Releasing the queue object")
            PDFCreatorQueue.ReleaseCom()
        End Try

    End Sub

    Private Sub mergedFiles_btn_Click(sender As Object, e As EventArgs) Handles mergedFiles_btn.Click
        Dim oJobQueue As Type = Type.GetTypeFromProgID("PDFCreator.JobQueue")
        Dim PDFCreatorQueue As Object = Activator.CreateInstance(oJobQueue)
        Dim PrintJob As Object
        Dim fullPath As String
        Const sampleName = "MergedMultipleFiles2Tif"

        Try
            fullPath = Path.Combine(Path.GetTempPath, sampleName + ".tif")

            MsgBox("Initializing PDFCreator queue...")
            PDFCreatorQueue.Initialize()

            MsgBox("Printing 3 windows test pages...")
            printFile()
            printFile()
            printFile()

            If Not PDFCreatorQueue.WaitForJobs(3, 15) Then
                MsgBox("The print jobs did not reach the queue within 15 seconds")
            Else
                MsgBox("Currently there are " & PDFCreatorQueue.Count & " job(s) in the queue")
                MsgBox("Merging all available jobs")
                PDFCreatorQueue.MergeAllJobs()

                MsgBox("Now there are " & PDFCreatorQueue.Count & " job(s) in the queue")
                MsgBox("Getting job instance")
                PrintJob = PDFCreatorQueue.NextJob

                PrintJob.SetProfileByGuidOrName("DefaultGuid")

                'We change the output format of the DefaultGuid to Tif only for these jobs
                MsgBox("Setting OutputFormat to tif")
                PrintJob.SetProfileSetting("OutputFormat", "Tif")

                MsgBox("Converting under ""DefaultGuid"" conversion profile but with .Tif as output format")
                PrintJob.ConvertTo(fullPath)

                If (Not PrintJob.IsFinished Or Not PrintJob.IsSuccessful) Then
                    MsgBox("Could not convert the file.")
                Else
                    MsgBox("Job finished successfully." + Environment.NewLine + Environment.NewLine + "File saved to: " & fullPath)
                End If
            End If
        Catch Ex As Exception
            MessageBox.Show("Original error: " & Ex.Message)
        Finally
            MsgBox("Releasing the queue object")
            PDFCreatorQueue.ReleaseCom()
        End Try
        ' End If
    End Sub

    'In order to keep the code small and clear, we'll only print the windows test page.
    'But you could also use the PDFCreatorObj object to print any file that you want.
    'Just select your file, pass the complete file path to PDFCreatorObj.PrintFile(filePath)
    'and it will be ready for conversion.
    Private Sub printFile()
        Dim ShellObj As Object

        ShellObj = CreateObject("Shell.Application")
        ShellObj.ShellExecute("RUNDLL32.exe", "PRINTUI.DLL,PrintUIEntry /k /n ""PDFCreator""", "", "open", 1)
    End Sub

End Class